move/sources/dirichlet.move

```diff
@@ -3,44 +3,60 @@ module addr::dirichlet {
 
-    ///     Sequential Dirichlet construction using Gamma variables.
-    ///     Each draw is independent and normalized at the end.
-    ///     Key insight: If X_1,...,X_n ~ Gamma(alpha_i, 1) independently,
-    ///     then (X_1/S,...,X_n/S) ~ Dirichlet(alpha_1,...,alpha_n) where S = sum(X_i)
     public(friend) fun sequential_dirichlet_hongbao(
-        total_amount: u64, num_packets: u64
-    ): FixedPoint32 {
-        // Use exponential distribution (Gamma with alpha=1) for simplicity
-        let draws: vector<FixedPoint32> = vector::empty();
-        let running_sum: FixedPoint32 = fixed_point32::create_from_u64(0);
-
-        // Phase 1: Sequential gamma draws
-        for (i in 0..num_packets) {
-            // Exponential is just Gamma(1,1)
-            let random = randomness::u32_integer();
-            // Scale random down by max u32, so it's 0->1
-            let random =
-                fixed_point32::create_from_rational((random as u64), (MAX_U32 as u64));
-
-            let draw = generate_exponential(random);
-            draws.push_back(draw);
-            running_sum = fixed_point32::create_from_raw_value(
-                running_sum.get_raw_value() + draw.get_raw_value()
+        remaining_amount: u64, remaining_packets: u64
+    ): u64 {
+        // Use stick breaking construction - Beta(1,a) where a is remaining_packets
+        // For each draw, take Beta(1,a) portion of remaining
+
+        // Convert random to FixedPoint32 between 0 and 1
+        let u =
+            fixed_point32::create_from_rational(
+                (randomness::u32_integer() as u64), 0xFFFFFFFF
+            );
+
+        let current_exp = {
+            let raw_exp = generate_exponential(u);
+            if (remaining_packets == 2) {
+                // Only adjust for the second-to-last position
+                let scale = fixed_point32::create_from_rational(9, 10); // 0.9
+                math_fixed::mul_div(fixed_point32::create_from_u64(1), raw_exp, scale)
+            } else {
+                raw_exp
+            }
+        };
+
+        // Use n-1 for remaining positions as it worked well for most positions
+        let remaining_exp = fixed_point32::create_from_u64(remaining_packets - 1);
+
+        // Calculate proportion
+        let total =
+            fixed_point32::create_from_raw_value(
+                current_exp.get_raw_value() + remaining_exp.get_raw_value()
+            );
+        let proportion =
+            math_fixed::mul_div(
+                fixed_point32::create_from_u64(1u64), current_exp, total
             );
+
+        // Calculate amount
+        let amount = fixed_point32::multiply_u64(remaining_amount, proportion);
+
+        // Ensure at least 1 token per packet
+        if (amount >= remaining_amount - (remaining_packets - 1)) {
+            amount = remaining_amount - (remaining_packets - 1);
+        };
+        if (amount == 0) {
+            amount = 1;
```

Comment:
It's possible to get here but the remaining balance is zero. That's fine if this function works that way but we should document it and short circuit in the balance = 0 case before we ever get to this function.